iwyu_ast_util.cc

```diff
@@ -1183,6 +1185,17 @@ bool IsBuiltinFunction(const clang::NamedDecl* decl) {
   return false;
 }
 
+bool IsImplicitlyInstantiated(const clang::FunctionDecl* decl) {
+  const FunctionTemplateSpecializationInfo* tpl_spec_info =
+      decl->getTemplateSpecializationInfo();
+  if (!tpl_spec_info)
+    return false;  // Not a template specialization.
+  const bool is_implicit = tpl_spec_info->getTemplateSpecializationKind() ==
+                           clang::TSK_ImplicitInstantiation;
+  const bool is_def = decl->isThisDeclarationADefinition();
```

Comment:
You could make this a guard clause up-front now, right?
```
if (!decl->isThisDeclarationADefinition())
  return false;
```
Not sure which is cheaper to test of `decl->getTemplateSpecializationInfo() == null` or `isThisDeclarationADefinition`, but you should be able to do them in that order.

Reply:
Maybe, just
```cpp
  return decl->isThisDeclarationADefinition() &&
         tpl_spec_info->getTemplateSpecializationKind() ==
             clang::TSK_ImplicitInstantiation;
```
?
Or the formatting is weird?

Reply:
I think that looks pretty good, works for me!